src/main/java/pl/javastart/couponcalc/PriceCalculator.java

```diff
@@ -1,13 +1,75 @@
 package pl.javastart.couponcalc;
 
+import java.math.BigDecimal;
+import java.math.RoundingMode;
 import java.util.List;
 
 public class PriceCalculator {
 
     public double calculatePrice(List<Product> products, List<Coupon> coupons) {
-        // TODO
+        double sum = 0;
+        if (coupons == null && products == null) {
+            return 0.0;
+        }
+        if (coupons == null) {
+            return getSumPriceWithoutDisc(products);
+        }
+        if (coupons.size() == 1) {
+            Coupon coupon = coupons.get(0);
+            if (!checkIfCouponHasCategory(coupon)) {
+                return getSumPriceWithCouponForAll(products, coupon);
+            } else {
+                return getSumPriceWithCouponWithCat(products, coupon);
+            }
+        } else {
+            double sumWithoutDisc = getSumPriceWithoutDisc(products);
+            double theBestDiff = 0;
+            for (Coupon coupon : coupons) {
+                double sumWithDisc = getSumPriceWithCouponWithCat(products, coupon);
+                double tempDiff = sumWithoutDisc - sumWithDisc;
+                if (tempDiff > theBestDiff) {
+                    theBestDiff = tempDiff;
+                }
```

Comment:
Wydaje mi się, że niepotrzebnie tutaj obliczasz cenę bez zniżki, a później liczysz różnicę  `sumWithoutDisc - sumWithDisc`.
Im mniejsza cena z zaaplikowanym kuponem, czyli `sumWithDisc`, tym lepiej i tak naprawdę to najmniejsza wartość `sumWithDisc ` powinna być wynikiem Twojej metody.